crates/telio-firewall/src/firewall.rs

```diff
@@ -476,29 +509,43 @@ impl StatefullFirewall {
             return false;
         }
 
-        match ip.get_next_level_protocol() {
-            IpNextHeaderProtocols::Udp => self.handle_inbound_udp(&whitelist, &peer, &ip),
-            IpNextHeaderProtocols::Tcp => self.handle_inbound_tcp(
-                |port| whitelist.is_port_whitelisted(&peer, port),
-                peer,
-                &ip,
-            ),
+        let check_port = |port| whitelist.is_port_whitelisted(&peer, port);
+
+        match proto {
+            IpNextHeaderProtocols::Udp => self.handle_inbound_udp(check_port, &peer, &ip),
+            IpNextHeaderProtocols::Tcp => self.handle_inbound_tcp(check_port, peer, &ip),
             IpNextHeaderProtocols::Icmp => self.handle_inbound_icmp(peer, &ip),
             IpNextHeaderProtocols::Icmpv6 if self.allow_ipv6 => self.handle_inbound_icmp(peer, &ip),
             _ => false,
         }
     }
 
-    fn handle_outbound_udp<'a>(&self, ip: &impl IpPacket<'a>) {
-        let key = unwrap_option_or_return!(Self::build_udp_key(ip, false));
+    fn handle_outbound_udp<'a>(&self, pubkey: PublicKey, ip: &impl IpPacket<'a>) {
+        let link = unwrap_option_or_return!(Self::build_udp_key(ip, false));
+        let key = UdpConn { link, pubkey };
+
+        let Some(headers) = ip
+            .packet()
+            .chunks_exact(UdpConnectionInfo::LAST_PKG_HEADER)
```

Comment:
If I understand correctly, this may fail for IP packets containing options, as then the IP header will be longer, and UDP header would get cut out?

refering to [IPv4Packet::get_header_length()](https://docs.rs/pnet/latest/pnet/packet/ipv4/struct.Ipv4Packet.html#method.get_header_length) could figure out the length of the header. and then adding 8 bytes as per [ICMP RFC](https://datatracker.ietf.org/doc/html/rfc792) probably would resolve the issue.

Reply:
I wanted to keep things on a stack array for performance. However, it seems to me that the length should be variable. In that case, let me store much more of a packet, I believe one of the RFCs recommends attaching the whole message to the ICMP packet

Reply:
Done, I've used the max IP header len (60) + UDP header len (8)